signer/jws.go

```diff
@@ -0,0 +1,480 @@
+	}
+
+	var crit []string
+	if val, ok := pHeaders[headerKeyCrit]; ok {
+		critical := reflect.ValueOf(val)
+		for i := 0; i < critical.Len(); i++ {
+			var val = critical.Index(i).Interface().(string)
+			if _, ok := headersMarkedCrit[val]; ok {
+				delete(headersMarkedCrit, val)
+			} else {
+				crit = append(crit, val)
+			}
+		}
+
+		if len(headersMarkedCrit) != 0 {
+			// This is not taken care by VerifySignerInfo method
+			return crit, MalformedSignatureError{"required headers not marked critical"}
+		}
+		return crit, nil
+	} else {
+		// This is not taken care by VerifySignerInfo method
+		return crit, MalformedSignatureError{"missing `crit` header"}
+	}
+}
+
+func getSignedAttrs(req SignRequest, method jwt.SigningMethod) map[string]interface{} {
+	attrs := make(map[string]interface{})
+	attrs[headerKeyAlg] = method.Alg()
+	attrs[headerKeySigningTime] = req.SigningTime.Format(time.RFC3339)
+	var crit = []string{headerKeySigningTime}
+	if !req.Expiry.IsZero() {
+		attrs[headerKeyExpiry] = req.Expiry.Format(time.RFC3339)
+		crit = append(crit, headerKeyExpiry)
+	}
+	attrs[headerKeyCty] = req.PayloadContentType
+
+	for _, elm := range req.ExtendedSignedAttrs {
+		attrs[elm.Key] = elm.Value
+		if elm.Critical {
+			crit = append(crit, elm.Key)
+		}
+	}
+
+	attrs[headerKeyCrit] = crit
+	return attrs
+}
+
+// ***********************************************************************
+// jwsEnvelope-JSON specific code
+// ***********************************************************************
+const (
+	// JWSPayloadContentType describes the media type of the jwsEnvelope envelope.
+	JWSPayloadContentType = "application/vnd.cncf.notary.v2.jws.v1"
+)
+
+// jwsInternalEnvelope is the final Signature envelope.
+type jwsInternalEnvelope struct {
+	// JWSPayload Base64URL-encoded.
+	Payload string `json:"Payload"`
+
+	// jwsProtectedHeader Base64URL-encoded.
+
+	Protected string `json:"Protected"`
+
+	// Signature metadata that is not integrity Protected
+	Header jwsUnprotectedHeader `json:"Header"`
+
+	// Base64URL-encoded Signature.
+	Signature string `json:"Signature"`
+}
+
+type jwsUnprotectedHeader struct {
+	// RFC3161 time stamp token Base64-encoded.
+	TimestampSignature []byte `json:"io.cncf.notary.TimestampSignature,omitempty"`
+
+	// List of X.509 Base64-DER-encoded certificates
+	// as defined at https://datatracker.ietf.org/doc/html/rfc7515#section-4.1.6.
+	CertChain [][]byte `json:"x5c"`
+
+	// SigningAgent used for signing
+	SigningAgent string `json:"io.cncf.notary.SigningAgent,omitempty"`
+}
+
+func newJwsInternalEnvelopeFromBytes(b []byte) (*jwsInternalEnvelope, error) {
+	var jws jwsInternalEnvelope
+	if err := json.Unmarshal(b, &jws); err != nil {
+		return &jws, err
+	}
+	return &jws, nil
+}
+
+
+func generateJws(compact string, req SignRequest) (*jwsInternalEnvelope, error) {
+	parts := strings.Split(compact, ".")
+	if len(parts) != 3 {
+		// this should never happen
+		return nil, errors.New("invalid compact serialization")
```

Comment:
> invalid compact serialization

This can be more descriptive. May be `unexpected error while generating a JWS signature envelope`?